blogpost/blogpost.md

```diff
@@ -0,0 +1,845 @@
+
+It wouldn't be complete without the `pipeline.run()` statement that actually executes our pipeline. As this is a streaming pipeline I don't use `.waitUntilFinish()` but instead return the `PipelineResult`. It will be easier to test.
+
+`.runPipeline()` is executed asynchronously right after the Bean's initialization in with `@PostConstruct` annotation's help.
+
+## LightningReceiver - read from Apache Kafka and write to MongoDB
+
+```java
+@Component
+@Log4j2
+@RequiredArgsConstructor
+public class LightningReceiver {
+  private final Environment env;
+  private final PipelineOptions pipelineOptions;
+  private final Executor executor;
+
+  @PostConstruct
+  public void runPipeline() {
+    executor.execute(this::lightningStreaming);
+  }
+
+  public PipelineResult lightningStreaming() {
+    Pipeline pipeline = Pipeline.create(pipelineOptions);
+
+    KafkaIO.Read<Long, String> kafkaRead = KafkaIO.<Long, String>read()
+        .withBootstrapServers(env.getProperty("kafka.bootstrap.servers"))
+        .withTopic(env.getProperty("kafka.topic"))
+        .withConsumerConfigUpdates(ImmutableMap.of("auto.offset.reset", env.getProperty("kafka.auto.offset.reset")))
+        .withKeyDeserializer(LongDeserializer.class)
+        .withValueDeserializer(StringDeserializer.class);
+
+    Integer limit = env.getProperty("kafka.limit", Integer.class);
+    if (limit != null) {
+      kafkaRead = kafkaRead.withMaxNumRecords(limit);
+    }
+
+    PCollection<KafkaRecord<Long, String>> kafkaRecords = pipeline.apply("Read data from Kafka", kafkaRead);
+
+    PCollection<Document> jsonLightningData = kafkaRecords.apply(
+        "Parse Kafka record to Mongo Document",
+        MapElements.into(TypeDescriptor.of(Document.class))
+            .via(kafkaRecord -> {
+              log.info("Read lightning {} from kafka", kafkaRecord.getKV().getValue());
+              return Document.parse(kafkaRecord.getKV().getValue());
+            }));
+
+    // Add separate branch to the pipeline tree
+    applyCountStrokeTheGroundLightnings(jsonLightningData);
+
+    jsonLightningData.apply("Write lightning data to MongoDB",
+        MongoDbIO.write()
+        .withDatabase(env.getProperty("mongo.database"))
+        .withCollection(env.getProperty("mongo.collection.lightnings"))
+        .withUri(env.getProperty("mongo.host")));
+
+    return pipeline.run();
+  }
+
+  private void applyCountStrokeTheGroundLightnings(PCollection<Document> pc) {
+    Integer windowSize = env.getProperty("beam.window.size", Integer.class);
+    pc
+      .apply("Filter hit the ground ones", Filter.by(document -> document.getBoolean("strokeTheGround")))
+      .apply("Add timestamp", WithTimestamps.of(document -> Instant.now()))
+      .apply("Apply fixed windows", Window.into(FixedWindows.of(Duration.standardSeconds(windowSize))))
+      .apply(
+          "Count lightnings that stroke the ground this minute",
```

Comment:
windowSize jest konfigurowalny, więc nie musi to być "this minute" :P 

Poza tym, z tego co widzę używasz processing time. A wtedy zdanie "Count lightnings that stroke the ground this minute" może być mylące, jeśli mamy opóźnienie i processing time znacząco odbiega od event time.

Możesz albo:
* wyraźnie zaznaczyć, że używasz processing time i liczysz te wyładowania, które Twoja aplikacja odczytała w ostatniej minucie,
* użyć event time. Tu jest wiele możliwości, można umieścić timestamp w wiadomości wysyłanej do Kafki, można brać pod uwagę czas otrzymania wiadomości przez Kafkę, itd.

Reply:
Dzięki! Myślę że już powinno być ok :)